paymentapi/src/main/java/fi/hel/verkkokauppa/payment/service/refund/PaytrailRefundPaymentService.java

```diff
@@ -0,0 +1,343 @@
+        if (!"confirmed".equals(refundStatus)) {
+            log.warn("creating refund payment for unconfirmed refund rejected, refundId: " + refundId);
+            throw new CommonApiException(
+                    HttpStatus.FORBIDDEN,
+                    new Error("rejected-creating-refund-payment-for-unconfirmed-refund", "rejected creating refund payment for unconfirmed refund, refundId id [" + refundId + "]")
+            );
+        }
+    }
+
+    public void setPaymentStatus(String paymentId, String status) {
+        RefundPayment payment = getRefund(paymentId);
+        payment.setStatus(status);
+        refundPaymentRepository.save(payment);
+    }
+
+    public RefundPayment getRefundPaymentForOrder(String orderId) {
+        List<RefundPayment> payments = refundPaymentRepository.findByOrderId(orderId);
+
+        RefundPayment paidPayment = selectPaidRefundPayment(payments);
+        RefundPayment payablePayment = selectPayableRefundPayment(payments);
+
+        if (paidPayment != null) {
+            return paidPayment;
+        } else if (payablePayment != null) {
+            return payablePayment;
+        } else {
+            return null;
+        }
+    }
+
+    public List<RefundPayment> getPaymentsForOrder(String orderId, String namepace) {
+        return refundPaymentRepository.findByNamespaceAndOrderId(namepace, orderId);
+    }
+
+    public List<RefundPayment> getPaymentsWithNamespaceAndOrderIdAndStatus(String orderId, String namepace, String status) {
+        return refundPaymentRepository.findByNamespaceAndOrderIdAndStatus(namepace, orderId, status);
+    }
+
+
+    private RefundPayment selectPayableRefundPayment(List<RefundPayment> payments) {
+        RefundPayment payablePayment = null;
+
+        if (payments != null)
+            for (RefundPayment payment : payments) {
+                // in an unpayable state
+                if (Objects.equals(payment.getStatus(), RefundPaymentStatus.PAID_ONLINE) || Objects.equals(payment.getStatus(), RefundPaymentStatus.CANCELLED)) {
+                    continue;
+                }
+
+                // an earlier selected refund payment is newer
+                if (payablePayment != null && payablePayment.getTimestamp().compareTo(payment.getTimestamp()) > 0) {
+                    continue;
+                }
+
+                payablePayment = payment;
+            }
+
+        return payablePayment;
+    }
+
+    private RefundPayment selectPaidRefundPayment(List<RefundPayment> payments) {
+        if (payments != null) {
+            for (RefundPayment payment : payments) {
+                if (Objects.equals(payment.getStatus(), RefundPaymentStatus.PAID_ONLINE)) {
+                    return payment;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    public RefundPayment getRefund(String refundId) {
+        Optional<RefundPayment> payment = refundPaymentRepository.findById(refundId);
```

Comment:
Maybe this method could be refactored to leverage the optional class to be more compact (formatting does not seem to work here, sorry for that...):

> return refundPaymentRepository.findById(refundId).orElseThrow(() -> {
>                  log.debug("refund not found, refundId: " + refundId);
>                  return new CommonApiException(
>                                 HttpStatus.NOT_FOUND,
>                                 new Error("refund-not-found-from-backend", "refund with refund id [" + refundId + "] not found from backend")
> }));

Reply:
These were just copy pasted code from payment, I will check these